src/compiler/crystal/syntax/parser.cr

```diff
@@ -5976,23 +5968,28 @@ module Crystal
       @token.to_s
     end
 
-    def push_def
-      @def_vars.push(Set(String).new)
-    end
-
-    def pop_def
-      @def_vars.pop
-    end
+    # If *create_scope* is true, creates an isolated variable scope and returns
+    # the yield result, resetting the scope afterwards. Otherwise simply returns
+    # the yield result without touching the scopes.
+    def with_isolated_var_scope(create_scope = true)
+      return yield unless create_scope
 
-    def push_def(args)
-      push_def(Set.new(args.map &.name))
-      ret = yield
-      pop_def
-      ret
+      begin
+        @var_scopes.push(Set(String).new)
+        yield
+      ensure
+        @var_scopes.pop
+      end
```

Comment:
```suggestion
      @var_scopes.push(Set(String).new)
      yield
    ensure
      @var_scopes.pop
```

Reply:
That would incorrectly pop for the early return branch.